### ISSUE TITLE
FYI: Sucessfully converted NanoDet to TensorFlow with eval implemetation in JavaScript

### ISSUE THREAD

**Issue by vladmandic** [RangiLyu/nanodet#187]
In case anybody is interested, NanoDet works like a charm in TensorFlow and TensorFlow/JS  

Full JS code as well as conversion notes are at:  
<https://gist.github.com/vladmandic/d0115e3899860391382d58550bd49b3f>

I've re-implemented score & box decoding to squeeze a bit more performance and moved NMS as a post-decoding task.

Posted example is for NodeJS, but the model equally works in browser using WebGL or WASM backends

Thank you for a great **TINY** model!

**Comment by batrlatom** [comment]
I have tried it, works great! I was trying to export the transformer version, I have succeeded. But the inference code is not compatible with such version. Do you think you would take a look on this? I think that I am missing something there. For ease, I am sharing already converted model ... https://recall-models.s3.eu-central-1.amazonaws.com/nanodet_transformer/model.json

**Comment by vladmandic** [comment]
it works with m and g variations, but not with t (transformer).
there is one trivial change needed:
```js
const baseSize = strideSize * 13;
```
should be `* 10` as transformer pretrained model has different convolution sizes  

but then you have the critical error:
```log
Message: In[0] and In[1] has different ndims: [400,1,128] vs. [128,384]
```
looking at the model code, this happens when two tensors passed to `matMul` op are not compatibile - something went wrong during the conversion  

i'd need to go over entire model workflow to figure out why (likely an incompatible broadcast, but that's just a guess), but at the end, seems like converter just cannot handle transformer model

**Comment by batrlatom** [comment]
Ok, thank you for checking it out. If you will have time sometimes and will try to use adapt it for transformer, just update it here, so we also will be able to use it. I tried the transformer network in the colab and it seems like great update over normal nanodet (320x320 resolution). It was able to pick a person inside the bus, I was surprised. Great work with the conversion then!!!

**Comment by batrlatom** [comment]
@vladmandic btw. I was playing with the code a little more and I am unable to get results that are similar to the results before the conversion. I am using the same conversion code as you described in the gist and the same code for inference. But results are different. Would you mind to share converted tfjs model for basic mobiledet_m_416 ?

**Comment by vladmandic** [comment]
@batrlatom here's a link, i'll keep it up for few days:  
<https://1drv.ms/u/s!ArSgEgBb24AcnMJcgIY1iwsA52WSWg?e=8TFmBa>

**Comment by batrlatom** [comment]
thanks, I will take a look!

**Comment by batrlatom** [comment]
There is not any problem with the conversion as your model gives me the same result. But the difference between js and vanilla implementation lies in the scores. Scores calculated by your code are about half the value of the original implementation. 

original output:
[https://recall-models.s3.eu-central-1.amazonaws.com/nanodet_416/test_original.png](url)
output from js:
[https://recall-models.s3.eu-central-1.amazonaws.com/nanodet_416/nanodet_js.png](url)

Nevertheless, the boxes are on similar spot, everything is fine, only scores are more off

**Comment by vladmandic** [comment]
i've decided to skip `softmax` op on scores output tensor, so it's by choice  

if you want, you can add it just before
```
const scoreIdx = scores.argMax(1).dataSync(); // location of highest scores
```

**Comment by batrlatom** [comment]
I get it. Thank you for the help and great work!

**Comment by zye1996** [comment]
unable to produce the result the same as the original, wondering if you have any idea? Thx

![k2-_372f1b5d-2b35-4684-a2a4-11da05cec113 v1-nanodet](https://user-images.githubusercontent.com/28901953/112079814-30148980-8b57-11eb-8552-8cded6ce0481.jpg)
![image](https://user-images.githubusercontent.com/28901953/112079882-51757580-8b57-11eb-804f-32b678a3e8f6.png)

**Comment by vladmandic** [comment]
probably i do have a bug - i'll take a look tomorrow.  
most likely it's due to different stride sizes require different score normalizations and i skipped that part.  

can you upload the original of the same photo so i can directly compare results in tests?

**Comment by zye1996** [comment]
sure no problem. here is the image

![](trash:///k2-_372f1b5d-2b35-4684-a2a4-11da05cec113.v1.jpg)

**Comment by vladmandic** [comment]
I've created a git repository as it's easier to update work than gist  
(plus i can upload actual model files):  
<https://github.com/vladmandic/nanodet>

The issue was that I was looking at only higest score within any given stride  
I've fixed that and now results are directly comparable  

Also, implementation is now directly aplicable to both `nanodet-m` and `nanodet-g` and it auto-adjusts  
(`nanodet-t` cannot be converted just yet)

To fine-tune results, modify `nanodet.js`: `modelOptions`  
Specifically, values for `minScore` and `iouThreshold`

**Comment by batrlatom** [comment]
I have tried your new code for nanodet.js with custom model but without success ( old code somehow works ) . Do you think you could take a look on it? For purpose of testing, the model is at : https://recall-models.s3.eu-central-1.amazonaws.com/nanodet_hardhat/model.json . There are 3 classes and you can take a image for prediction for example at https://images.unsplash.com/photo-1603516270950-26e4f5004ffd?ixid=MXwxMjA3fDB8MHxzZWFyY2h8Mnx8aGFyZGhhdHxlbnwwfHwwfA%3D%3D&ixlib=rb-1.2.1&w=1000&q=80

**Comment by vladmandic** [comment]
@batrlatom 

my implementation had a hard-coded number of expected classes to 80  
as that is number in coco classes which is what original model has been trained on:  

```js
  const scoresT = res.find((a) => (a.shape[1] === (baseSize ** 2) && a.shape[2] === 80))?.squeeze();
```

and your custom model has only 3 classes, as it can be seen from the `model.signature.output`:
```json
'Identity:0': { name: 'Identity:0', dtype: 'DT_FLOAT', tensorShape: { dim: [ { size: '1' }, { size: '2704' }, { size: '3' } } },
```
so my implementation could not find correct tensor that represents scores

I just changed hard-coded value to value of labels.length,  
so as long as you're importing labels that match the model you should be fine:  
```js
const { labels } = require('./coco-labels');
```

note: just make sure your number of classes is never 32 or 44 as that is dimension  
of boxes tensor and there is nothing else to base tensor match operation on :)

**Comment by batrlatom** [comment]
Works well, thank you!

**Comment by Techn08055** [comment]
> In case anybody is interested, NanoDet works like a charm in TensorFlow and TensorFlow/JS
> 
> Converted models, full JS code as well as conversion notes are at: https://github.com/vladmandic/nanodet
> 
> I've re-implemented score & box decoding to squeeze a bit more performance and moved NMS as a post-decoding task to avoid calculation of low-probability boxes (code is fully commented with explanations)
> 
> Implementation is now directly applicable to both `nanodet-m` and `nanodet-g` and it auto-adjusts (`nanodet-t` converted model has some corrupted layers and cannot be used just yet)
> 
> Model works in NodeJS using `TensorFlow` backend and Browser using `WebGL` backend, but does not work using `WASM` backend due to missing implementation for an operation `SparseToDense` ([tensorflow/tfjs#4824](https://github.com/tensorflow/tfjs/issues/4824))
> 
> Thank you for a great **TINY** model!


Hey I tried converting to tensorflow and tflite but the model seems to perform poor I tried the inference provided  here:
https://github.com/PINTO0309/PINTO_model_zoo/blob/main/072_NanoDet/demo/demo_tflite.py

**Comment by lijianyuhoopox** [comment]
looks like the mode not OK.

![1595001411699](https://user-images.githubusercontent.com/107986839/176865090-84973165-4423-4829-b8a2-818e2d650a62.jpeg)

**Comment by adithya1-7** [comment]
Hey @vladmandic
Can you help me with this issue
#460
